source/agora/consensus/state/ValidatorSet.d

```diff
@@ -138,11 +141,13 @@ public class ValidatorSet
             return "A validator with the same public key is already enrolled";
 
         () @trusted {
-            this.db.execute("INSERT OR REPLACE INTO preimages " ~
-                "(key, height, preimage) " ~
-                "VALUES (?, ?, ?)",
-                enroll.utxo_key, height.value,
-                enroll.commitment);
+            // Only add or update if we do not have a newer preimage already
+            if (this.getPreimage(enroll.utxo_key).height <= height)
+                this.db.execute("INSERT OR REPLACE INTO preimages " ~
+                    "(key, height, preimage) " ~
+                    "VALUES (?, ?, ?)",
+                    enroll.utxo_key, height.value,
+                    enroll.commitment);
```

Comment:
Wouldn't it be better to just call `addPreimage` here (after the next line though) ?

Reply:
In `addPreimage` it always does an update and not an insert. Do you mean update that function to handle both cases?

Reply:
Actually that is better as we can check the height for any pre-image added is more than stored height.

Reply:
Hum right, I missed that. I'm just looking at the statement and wondering if we could improve it be removing the conditional. But I haven't found a good approach, so let's go ahead with the current state.

Reply:
When we receive a `pre-mage` via a post in `FullNode` we don't currently check the height is higher so we need the check anyway.

Reply:
Just realized we already check in the `addPreimage` function that the height is higher than previous.